File: test/integration/harness/file_review.go
package harness

import (
	"context"
	"fmt"
	"testing"
	"time"

	agentexecv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agentexecution/v1"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// File-review (apply-then-review FileChangeSet HITL) test helpers.
//
// These are the file-review siblings of the approval helpers in
// agent_execution_waiter.go / approval_stream.go. A git-backed harness (Cursor
// deny-only or the native deep-agent) lets edits FLOW during a turn and reviews
// the captured baseline->candidate diff at the turn boundary: the runner authors
// a FileChangeSet (status AWAITING_REVIEW), the human keeps/discards each file
// via SubmitFileDecision, and the runner reconciles the approved bytes on resume.
// All of these helpers read the SERVER projection (status.file_change_sets) and
// the append-only audit ledger (status.file_review_event_stream) — never any
// runner-internal state — so they hold identically for both harnesses and both
// editions.

// WaitForFileReview polls until the execution has at least one change set
// awaiting a human decision (status AWAITING_REVIEW), or times out. It is the
// file-review counterpart of WaitForApproval: a turn that edited tracked files
// reaches EXECUTION_WAITING_FOR_APPROVAL with the captured set projected onto
// status.file_change_sets. Returns the terminal snapshot together with an error
// if the execution finishes before any set is offered for review (e.g. the LLM
// did not edit a file, or every edit landed on a gitignored path that stays
// true-paused instead of being captured).
func (w *AgentExecutionWaiter) WaitForFileReview(ctx context.Context, executionID string, timeout time.Duration) (*agentexecv1.AgentExecution, error) {
	if timeout == 0 {
		timeout = defaultTimeout
	}

	deadline := time.Now().Add(timeout)
	interval := defaultPollInterval

	for time.Now().Before(deadline) {
		exec, err := w.client.Get(ctx, &agentexecv1.AgentExecutionId{Value: executionID})
		if err != nil {
			w.logger.Debug("agent execution poll error (will retry)", "execution_id", executionID, "error", err)
			select {
			case <-ctx.Done():
				return nil, ctx.Err()
			case <-time.After(interval):
			}
			interval = nextInterval(interval)
			continue
		}

		if FindFileChangeSet(exec, agentexecv1.FileChangeSetStatus_FILE_CHANGE_SET_STATUS_AWAITING_REVIEW) != nil {
			return exec, nil
		}

		if isAgentTerminalPhase(exec.GetStatus().GetPhase()) {
			return exec, fmt.Errorf(
				"agent execution %s reached terminal phase %s before any change set was offered for review",
				executionID, exec.GetStatus().GetPhase().String())
		}

		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(interval):
		}
		interval = nextInterval(interval)
	}

	return nil, fmt.Errorf("timed out waiting for a file change set awaiting review on execution %s after %v",
		executionID, timeout)
}

// WaitForFileChangeProgress polls until the execution exposes a MID-RUN
// file_change_progress snapshot that satisfies pred — i.e. a CAPTURING change
// set exists AND status.file_change_progress is populated for it AND pred
// accepts it (DD-32 / DD-33). It is the mid-run counterpart of WaitForFileReview:
// where that waits for the turn-boundary AWAITING_REVIEW set, this observes the
// transient pre-boundary window. Modeled on WaitForExecutionUsageReport — the
// same shape used elsewhere to poll an eventually-settled value instead of
// sleeping and hoping.
//
// The predicate is load-bearing, not a convenience. The runner attaches
// file_change_progress on the FIRST streaming persist — which fires before a
// file is written — so an early snapshot legitimately carries files_changed=0.
// Keying only on "progress != nil" could therefore return that pre-write
// snapshot and race the writes. Callers pass a settled predicate (e.g.
// `p.GetFilesChanged() == 2`) whose target is stable for the whole CAPTURING
// span, making the observation deterministic.
//
// The last-seen execution is returned even on timeout or an early terminal
// phase, so callers can run their assertions against it and produce a diagnostic
// failure (e.g. "progress settled at files_changed=1") rather than a bare
// timeout.
func (w *AgentExecutionWaiter) WaitForFileChangeProgress(
	ctx context.Context,
	executionID string,
	timeout time.Duration,
	pred func(*agentexecv1.FileChangeProgress) bool,
) (*agentexecv1.AgentExecution, error) {
	if timeout == 0 {
		timeout = defaultTimeout
	}

	deadline := time.Now().Add(timeout)
	interval := defaultPollInterval

	var last *agentexecv1.AgentExecution
	for time.Now().Before(deadline) {
		exec, err := w.client.Get(ctx, &agentexecv1.AgentExecutionId{Value: executionID})
		if err == nil {
			last = exec
			progress := exec.GetStatus().GetFileChangeProgress()
			capturing := FindFileChangeSet(exec, agentexecv1.FileChangeSetStatus_FILE_CHANGE_SET_STATUS_CAPTURING)
			if capturing != nil && progress != nil && (pred == nil || pred(progress)) {
				return exec, nil
			}

			if isAgentTerminalPhase(exec.GetStatus().GetPhase()) {
				return last, fmt.Errorf(
					"agent execution %s reached terminal phase %s before a mid-run file_change_progress snapshot satisfied the predicate",
					executionID, exec.GetStatus().GetPhase().String())
			}
		} else {
			w.logger.Debug("agent execution poll error (will retry)", "execution_id", executionID, "error", err)
		}

		select {
		case <-ctx.Done():
			return last, ctx.Err()
		case <-time.After(interval):
		}
		interval = nextInterval(interval)
	}

	return last, fmt.Errorf(
		"timed out after %v waiting for a mid-run file_change_progress snapshot on execution %s to satisfy the predicate",
		timeout, executionID)
}

// FindFileChangeSet returns the first projected change set in the given status,
// or nil if none match. Mirror of FindPendingApproval for the file-review
// projection (status.file_change_sets, recomputed from the file_review ledger on
// every status write).
func FindFileChangeSet(exec *agentexecv1.AgentExecution, status agentexecv1.FileChangeSetStatus) *agentexecv1.FileChangeSet {
	for _, cs := range exec.GetStatus().GetFileChangeSets() {
		if cs.GetStatus() == status {
			return cs
		}
	}
	return nil
}

// CountChangeSets returns how many projected change sets are in the given status.
func CountChangeSets(exec *agentexecv1.AgentExecution, status agentexecv1.FileChangeSetStatus) int {
	n := 0
	for _, cs := range exec.GetStatus().GetFileChangeSets() {
		if cs.GetStatus() == status {
			n++
		}
	}
	return n
}

// FindCapturedChangeByPath returns the captured change for the given
// workspace-relative path within a change set, or nil if none match. An
// ADD/MODIFY change is keyed by its after-path; a DELETE by its before-path
// (after-path is empty), so this matches either side.
func FindCapturedChangeByPath(set *agentexecv1.FileChangeSet, path string) *agentexecv1.CapturedFileChange {
	for _, ch := range set.GetChanges() {
		if ch.GetPathAfter() == path || ch.GetPathBefore() == path {
			return ch
		}
	}
	return nil
}

// AssertCapturedChange asserts that a change set carries a captured change for
// the given path with the expected kind, and returns it for further
// change-specific assertions (before/after content, sha256 digests, the
// file_digest the decision must echo as expected_digest). On absence it fails
// the test (non-fatally) and returns nil.
func AssertCapturedChange(
	t *testing.T,
	set *agentexecv1.FileChangeSet,
	path string,
	expectedKind agentexecv1.FileChangeKind,
) *agentexecv1.CapturedFileChange {
	t.Helper()
	ch := FindCapturedChangeByPath(set, path)
	if ch == nil {
		t.Errorf("expected a captured change for path %q in change set %q, found none (%d change(s) present)",
			path, set.GetId(), len(set.GetChanges()))
		return nil
	}
	assert.Equalf(t, expectedKind, ch.GetKind(),
		"captured change %q: expected kind %s, got %s",
		path, expectedKind.String(), ch.GetKind().String())
	assert.NotEmptyf(t, ch.GetFileDigest(),
		"captured change %q: file_digest must be populated (it is the expected_digest a FILE-scoped decision echoes)", path)
	return ch
}

// FileReviewStreamHasEvent reports whether the persisted, append-only
// file-review event stream contains an event of the given type for the given
// change set. It is the file-review sibling of ApprovalStreamHasEvent: the
// stream is the server-authored audit ledger that status.file_change_sets is
// projected from, so both suites assert against it through this one predicate.
func FileReviewStreamHasEvent(
	stream *agentexecv1.FileReviewEventStream,
	changeSetID string,
	eventType agentexecv1.FileReviewEventType,
) bool {
	for _, ev := range stream.GetEvents() {
		if ev.GetChangeSetId() == changeSetID && ev.GetEventType() == eventType {
			return true
		}
	}
	return false
}

// SubmitFileDecisionByPath records a FILE-scoped keep/discard decision on one
// captured change, looked up by its workspace-relative path. It resolves the
// required expected_digest from the projected change (the file_digest the
// reviewer saw), enforcing the "what you approve is what gets applied" contract
// the server re-checks. Fails the test fatally if the change is absent.
func SubmitFileDecisionByPath(
	t *testing.T,
	ctx context.Context,
	clients *Clients,
	executionID string,
	set *agentexecv1.FileChangeSet,
	path string,
	action agentexecv1.FileDecisionAction,
) {
	t.Helper()
	ch := FindCapturedChangeByPath(set, path)
	require.NotNilf(t, ch, "cannot decide path %q: no captured change in set %q", path, set.GetId())

	_, err := clients.AgentExecutionCommand.SubmitFileDecision(ctx, &agentexecv1.SubmitFileDecisionInput{
		AgentExecutionId: executionID,
		ChangeSetId:      set.GetId(),
		Scope:            agentexecv1.FileDecisionScope_FILE_DECISION_SCOPE_FILE,
		FileChangeId:     ch.GetId(),
		Action:           action,
		ExpectedDigest:   ch.GetFileDigest(),
	})
	require.NoErrorf(t, err, "submit FILE decision (%s) for %q in set %q", action.String(), path, set.GetId())
}

// SubmitChangeSetDecision records one CHANGE_SET-scoped decision that covers
// every change in the set in a single call (the "Keep all" / "Discard all"
// action). It echoes the set's aggregate_digest as expected_digest.
func SubmitChangeSetDecision(
	t *testing.T,
	ctx context.Context,
	clients *Clients,
	executionID string,
	set *agentexecv1.FileChangeSet,
	action agentexecv1.FileDecisionAction,
) {
	t.Helper()
	_, err := clients.AgentExecutionCommand.SubmitFileDecision(ctx, &agentexecv1.SubmitFileDecisionInput{
		AgentExecutionId: executionID,
		ChangeSetId:      set.GetId(),
		Scope:            agentexecv1.FileDecisionScope_FILE_DECISION_SCOPE_CHANGE_SET,
		Action:           action,
		ExpectedDigest:   set.GetAggregateDigest(),
	})
	require.NoErrorf(t, err, "submit CHANGE_SET decision (%s) for set %q", action.String(), set.GetId())
}

// ResolveFileDecisionsByPathUntilPhase steps an execution through file review by
// deciding every captured change with a per-path callback, until the execution
// reaches target or times out. It is the file-review counterpart of
// ResolveApprovalsByPathUntilPhase.
//
// Unlike tool approvals, ordering does NOT matter here: SubmitFileDecision only
// signals the workflow to resume once the WHOLE gate clears (no change set
// awaiting review AND no pending approval — see submit_file_decision.go
// GateResolved). A file REJECT has no immediate-resume shortcut, so every change
// in the set must be decided before the runner reconciles. This submits a
// FILE-scoped decision for each change in each AWAITING_REVIEW set, then waits
// for the resume to drive the execution to target.
//
// decideByPath receives each captured change's path (after-path, or before-path
// for a delete) and returns APPROVE (keep) or REJECT (discard). Decided changes
// move the set out of AWAITING_REVIEW, so re-polling never re-submits.
func (w *AgentExecutionWaiter) ResolveFileDecisionsByPathUntilPhase(
	t *testing.T,
	ctx context.Context,
	clients *Clients,
	executionID string,
	decideByPath func(path string) agentexecv1.FileDecisionAction,
	target agentexecv1.ExecutionPhase,
	timeout time.Duration,
) (*agentexecv1.AgentExecution, error) {
	t.Helper()
	deadline := time.Now().Add(timeout)
	interval := 500 * time.Millisecond

	for time.Now().Before(deadline) {
		exec, err := w.client.Get(ctx, &agentexecv1.AgentExecutionId{Value: executionID})
		if err != nil {
			return nil, fmt.Errorf("get execution %s: %w", executionID, err)
		}

		phase := exec.GetStatus().GetPhase()
		if phase == target {
			return exec, nil
		}

		for _, set := range exec.GetStatus().GetFileChangeSets() {
			if set.GetStatus() != agentexecv1.FileChangeSetStatus_FILE_CHANGE_SET_STATUS_AWAITING_REVIEW {
				continue
			}
			for _, ch := range set.GetChanges() {
				path := ch.GetPathAfter()
				if path == "" {
					path = ch.GetPathBefore()
				}
				SubmitFileDecisionByPath(t, ctx, clients, executionID, set, path, decideByPath(path))
			}
		}

		if isAgentTerminalPhase(phase) && phase != target {
			return exec, fmt.Errorf(
				"agent execution reached terminal phase %s instead of expected %s",
				phase.String(), target.String(),
			)
		}

		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(interval):
		}
	}

	return nil, fmt.Errorf(
		"timed out waiting for agent execution %s to reach phase %s after %v",
		executionID, target.String(), timeout,
	)
}

File: test/conformance/src/support/file-review.ts
// File-review (apply-then-review HITL) helpers for the execution suites.
// Domain: conformance support (execution engine).
//
// A turn that edits files in a git workspace does not gate per tool call: the
// runner captures the whole delta at the turn boundary and the server projects
// it onto status.file_change_sets as a FileChangeSet AWAITING_REVIEW while the
// execution sits in EXECUTION_WAITING_FOR_APPROVAL with ZERO pending_approvals
// (the two gates are siblings, not one). Decisions go through
// submitFileDecision at FILE or CHANGE_SET scope, bound to the digest the
// reviewer saw; the runner reconciles the approved bytes and the ledger records
// RECONCILED. These helpers are the read-and-decide vocabulary for that
// lifecycle, the way support/agentexecutions.ts is for the approval gate.
//
// Correlation is ALWAYS by stable id (change set id, file change id); digests
// are the enforcement gate the decision carries, never how a change is found
// (filereview.proto's identity rule). So a decision "by path" here looks the
// change up by path_after/path_before and then submits ITS id and ITS digest.
//
// Ported from the retired Go harness's file_review.go; the wait semantics are
// its: a set AWAITING_REVIEW resolves the wait, and a terminal phase before one
// appears fails it with the phase.
//
// The mid-run progress wait is different in kind: file_change_progress is a
// display snapshot the runner captures on its own persist cadence, and a
// capture runs while the tool that triggered the persist may still be writing
// (the runner's shared/filereview/progress.ts header, "What a capture may
// see"). So one snapshot can be a turn behind, half-written, or a turn ahead,
// and only the CONVERGED shape is a contract: an arm states the entries it
// expects and waits until a CAPTURING snapshot carries all of them with their
// counts (`awaitProgressEntries`), never for "the first snapshot with N
// files". Every progress wait records the distinct snapshots it saw and puts
// them in its failure message, because the sequence (say 0 -> 1 -> 3) is the
// diagnosis a bare timeout hides.
import type { AgentExecution } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/api_pb";
import {
  FileChangeKind,
  FileChangeSetStatus,
  FileDecisionAction,
  FileDecisionScope,
  FileReviewEventType,
} from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/enum_pb";
import type {
  CapturedFileChange,
  FileChangeProgress,
  FileChangeProgressEntry,
  FileChangeSet,
} from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/filereview_pb";
import type { ConformanceClients } from "../harness/clients";
import { isTerminalPhase, pollExecution, type PollOptions } from "./agentexecutions";

// Polls until a change set is offered for review. Fails — with the phase — if
// the execution ends first (the agent edited nothing, or every edit landed on a
// path that stays true-paused), so a wrong fixture is a legible failure rather
// than a timeout.
export function awaitFileReview(
  clients: ConformanceClients,
  executionId: string,
  opts: PollOptions = {},
): Promise<AgentExecution> {
  return pollExecution(
    clients,
    executionId,
    (exec) => {
      if (findChangeSet(exec, FileChangeSetStatus.AWAITING_REVIEW) !== undefined) return true;
      if (isTerminalPhase(exec.status?.phase)) {
        throw new Error(
          `execution ${executionId} reached terminal phase before any change set was offered for review ` +
            `(status.error: ${JSON.stringify(exec.status?.error ?? "")})`,
        );
      }
      return false;
    },
    { label: "a change set AWAITING_REVIEW", ...opts },
  );
}

// One row an arm expects a mid-run progress snapshot to converge to: the path
// (path_after for ADD/MODIFY, path_before for DELETE), the kind and the exact
// line counts. Counts are part of the expectation on purpose: a snapshot
// captured while the tool was still writing carries the right path with the
// wrong counts, and waiting for the counts is what makes the wait land on a
// converged capture rather than a half-written one.
export interface ExpectedProgressEntry {
  readonly path: string;
  readonly kind: FileChangeKind;
  readonly linesAdded: number;
  readonly linesRemoved: number;
}

// The snapshot's row for a workspace-relative path, if present.
export function progressEntryFor(progress: FileChangeProgress, path: string): FileChangeProgressEntry | undefined {
  return progress.entries.find((e) => e.pathAfter === path || e.pathBefore === path);
}

// Whether `progress` carries EVERY expected row with exactly its kind and
// counts. Extra rows are allowed: a capture that lands after a later tool's
// write carries that tool's file too, and when a capture runs relative to the
// tool it rides with is not part of the contract (header).
export function progressCarries(progress: FileChangeProgress, expected: readonly ExpectedProgressEntry[]): boolean {
  return expected.every((want) => {
    const got = progressEntryFor(progress, want.path);
    return (
      got !== undefined &&
      got.kind === want.kind &&
      got.linesAdded === want.linesAdded &&
      got.linesRemoved === want.linesRemoved
    );
  });
}

// Polls until a MID-RUN file_change_progress snapshot satisfies `settled` while
// a CAPTURING set exists — the transient pre-boundary window. `settled` must
// hold for the converged snapshot and stay true once it does (header); an arm
// with a concrete expectation uses `awaitProgressEntries`. On failure — a
// timeout, or the boundary arriving first — the error carries the distinct
// snapshots observed, in order, with the time each was first seen.
export async function awaitFileChangeProgress(
  clients: ConformanceClients,
  executionId: string,
  settled: (progress: FileChangeProgress) => boolean,
  opts: PollOptions = {},
): Promise<AgentExecution> {
  const trace = new ProgressTrace();
  try {
    return await pollExecution(
      clients,
      executionId,
      (exec) => {
        const progress = exec.status?.fileChangeProgress;
        const capturing = findChangeSet(exec, FileChangeSetStatus.CAPTURING);
        trace.observe(progress);
        if (capturing !== undefined && progress !== undefined && settled(progress)) return true;
        if (isTerminalPhase(exec.status?.phase)) {
          throw new Error(
            `execution ${executionId} reached terminal phase before a mid-run file_change_progress snapshot ` +
              `satisfied the predicate`,
          );
        }
        return false;
      },
      { label: "a settled mid-run file_change_progress snapshot", ...opts },
    );
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    throw new Error(`${message}; ${trace.render()}`, { cause: err });
  }
}

// Polls until a CAPTURING snapshot carries every expected row (`progressCarries`).
// The shape every progress arm should take: state what the strip converges to,
// not how many files one capture happened to see.
export function awaitProgressEntries(
  clients: ConformanceClients,
  executionId: string,
  expected: readonly ExpectedProgressEntry[],
  opts: PollOptions = {},
): Promise<AgentExecution> {
  const label = `a mid-run file_change_progress snapshot carrying ${expected.map((e) => e.path).join(", ")}`;
  return awaitFileChangeProgress(clients, executionId, (p) => progressCarries(p, expected), { label, ...opts });
}

// The distinct progress snapshots a wait observed, rendered for a failure
// message: `observed progress: +1.0s files=1 [created.txt ADD +3/-0] -> +6.9s
// files=3 [...]`. Consecutive identical snapshots collapse to one entry, so the
// rendering is the sequence of captures, not of polls.
class ProgressTrace {
  private readonly startedAt = Date.now();
  private readonly seen: string[] = [];
  private last: string | undefined;

  observe(progress: FileChangeProgress | undefined): void {
    const rendered = progress === undefined ? "none" : renderProgress(progress);
    if (rendered === this.last) return;
    this.last = rendered;
    const elapsed = ((Date.now() - this.startedAt) / 1000).toFixed(1);
    this.seen.push(`+${elapsed}s ${rendered}`);
  }

  render(): string {
    return `observed progress: ${this.seen.length === 0 ? "(nothing)" : this.seen.join(" -> ")}`;
  }
}

function renderProgress(progress: FileChangeProgress): string {
  const rows = progress.entries.map(
    (e) => `${e.pathAfter || e.pathBefore} ${FileChangeKind[e.kind]} +${e.linesAdded}/-${e.linesRemoved}`,
  );
  return `files=${progress.filesChanged} [${rows.join(", ")}]`;
}

// The first projected change set in `status`, if any — the file-review analogue
// of scanning pending_approvals.
export function findChangeSet(exec: AgentExecution, status: FileChangeSetStatus): FileChangeSet | undefined {
  return exec.status?.fileChangeSets.find((set) => set.status === status);
}

// The set AWAITING_REVIEW, which every decide-and-continue arm starts from;
// throws when absent so an assertion never dereferences a missing set.
export function requireReviewSet(exec: AgentExecution): FileChangeSet {
  const set = findChangeSet(exec, FileChangeSetStatus.AWAITING_REVIEW);
  if (set === undefined) {
    throw new Error(
      `execution ${exec.metadata?.id ?? "?"} has no change set AWAITING_REVIEW ` +
        `(sets: ${JSON.stringify(exec.status?.fileChangeSets.map((s) => FileChangeSetStatus[s.status]))})`,
    );
  }
  return set;
}

// The change for a workspace-relative path: path_after for ADD/MODIFY/RENAME,
// path_before for DELETE. Throws when absent so a typo in an arm is a named
// failure, not an undefined dereference.
export function requireChangeByPath(set: FileChangeSet, path: string): CapturedFileChange {
  const change = set.changes.find((c) => c.pathAfter === path || c.pathBefore === path);
  if (change === undefined) {
    throw new Error(
      `change set ${set.id} has no change for ${path} ` +
        `(paths: ${JSON.stringify(set.changes.map((c) => c.pathAfter || c.pathBefore))})`,
    );
  }
  return change;
}

export interface FileDecisionOptions {
  // Overrides the digest the decision is bound to. The default is the target's
  // CURRENT digest (what the reviewer saw); an arm passes a stale value to
  // prove the enforcement gate refuses it.
  expectedDigest?: string;
  // Consciously keep a binary change whose diff cannot be shown (APPROVE at
  // FILE scope only, per SubmitFileDecisionInput).
  acknowledgeUnreviewable?: boolean;
  reason?: string;
}

// Decides one file, correlating by the change's id and binding to its
// file_digest. Returns the server's response (the execution as the handler
// left it) so an arm can assert the projection synchronously.
export function submitFileDecisionByPath(
  clients: ConformanceClients,
  executionId: string,
  set: FileChangeSet,
  path: string,
  action: FileDecisionAction,
  opts: FileDecisionOptions = {},
): Promise<AgentExecution> {
  const change = requireChangeByPath(set, path);
  return clients.agentExecutionCommand.submitFileDecision({
    agentExecutionId: executionId,
    changeSetId: set.id,
    scope: FileDecisionScope.FILE,
    fileChangeId: change.id,
    action,
    expectedDigest: opts.expectedDigest ?? change.fileDigest,
    acknowledgeUnreviewable: opts.acknowledgeUnreviewable ?? false,
    reason: opts.reason ?? "",
  });
}

// Decides the whole set, bound to its aggregate_digest.
export function submitChangeSetDecision(
  clients: ConformanceClients,
  executionId: string,
  set: FileChangeSet,
  action: FileDecisionAction,
  opts: FileDecisionOptions = {},
): Promise<AgentExecution> {
  return clients.agentExecutionCommand.submitFileDecision({
    agentExecutionId: executionId,
    changeSetId: set.id,
    scope: FileDecisionScope.CHANGE_SET,
    action,
    expectedDigest: opts.expectedDigest ?? set.aggregateDigest,
    acknowledgeUnreviewable: opts.acknowledgeUnreviewable ?? false,
    reason: opts.reason ?? "",
  });
}

// Whether the file-review ledger records `type` for `changeSetId` (any set when
// omitted). The ledger is the source of truth the projection is derived from,
// so RECONCILED here is the proof the runner applied the decision.
export function fileReviewStreamHas(
  exec: AgentExecution,
  type: FileReviewEventType,
  changeSetId?: string,
): boolean {
  return (exec.status?.fileReviewEventStream?.events ?? []).some(
    (event) => event.eventType === type && (changeSetId === undefined || event.changeSetId === changeSetId),
  );
}
